tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -348,24 +352,580 @@ void EnqueueWriteBufferCommand::process() {
+                flush.relay_inline.stride = align(flush.relay_inline.stride, L1_ALIGNMENT);
+            }
+            // In general, length does not include alignment
+            flush.relay_inline.length = flush.relay_inline.stride - sizeof(CQPrefetchCmd);
+            flush.relay_inline.stride = align(flush.relay_inline.stride, HUGEPAGE_ALIGNMENT);
+
+            flush_ptr = (uint32_t*)&flush;
+            for (int i = 0; i < sizeof(CQPrefetchCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(*flush_ptr++);
+            }
+            CQDispatchCmd write_packed_cmd;
+            write_packed_cmd.base.cmd_id = CQ_DISPATCH_CMD_WRITE_PACKED;
+            write_packed_cmd.write_packed.is_multicast = false;
+            write_packed_cmd.write_packed.count = num_packed_cmds;
+            write_packed_cmd.write_packed.addr = 0;
+            write_packed_cmd.write_packed.size = 0;
+
+            uint32_t* write_packed_cmd_ptr = (uint32_t*)&write_packed_cmd;
+            for (int i = 0; i < sizeof(CQDispatchCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(*write_packed_cmd_ptr++);
+            }
+            for (int i = 0; i < num_packed_cmds * sizeof(CQDispatchWritePackedUnicastSubCmd) / sizeof(uint32_t); i++) {
+                this->commands.push_back(0);
+            }
+            if ((padding = (sizeof(CQPrefetchCmd) + sizeof(CQDispatchCmd) +
+                            num_packed_cmds * sizeof(CQDispatchWritePackedUnicastSubCmd)) %
+                           L1_ALIGNMENT) != 0) {
+                for (int j = 0; j < (L1_ALIGNMENT - padding) / sizeof(uint32_t); j++) {
+                    this->commands.push_back(0);
+                }
+            }
+            // Runtime Args Data
+            for (int i = 0; i < num_packed_cmds; i++) {
+                for (int j = 0; j < transfer_info[i].data.size(); j++) {
+                    this->commands.push_back(0);
+                }
+                if ((padding = (transfer_info[i].data.size() * sizeof(uint32_t)) % L1_ALIGNMENT) != 0) {
+                    for (int j = 0; j < (L1_ALIGNMENT - padding) / sizeof(uint32_t); j++) {
+                        this->commands.push_back(0);
+                    }
+                }
+            }
+            align_commands_vector(this->commands, HUGEPAGE_ALIGNMENT);
+        }
+        // Multicast Semaphores Cmd
+        for (const auto& [dst, transfer_info_vec] : program.program_transfer_info.multicast_semaphores) {
+            std::uint32_t num_packed_cmds = 0;
+            std::uint32_t write_packed_len = transfer_info_vec[0].data.size();
+            for (const auto& transfer_info: transfer_info_vec) {
+                for (const auto& dst_noc_info: transfer_info.dst_noc_info) {
+                    TT_ASSERT(transfer_info.data.size() == write_packed_len, "Not all data vectors in write packed semaphore cmd equal in len");
+                    num_packed_cmds += 1;
```

Comment:
these loops basically do a bunch of asserts and a multiply, right?

Reply:
Right.. I think for these write packed cmds, there is flexibility to potentially support different data sizes for, maybe runtime args.. If we can assert at the host API level that the data for each CoreRange is same len, this can be just a multiply